### PR TITLE
feat(system_monitor): check UDP network errors

### PR DESCRIPTION
**Pull request by iwatake2222**
### Why this  PR is needed

- UDP packet drops make Autoware unstable. Therefore, this is an important indicator for determining whether Autoware is functioning correctly
- There are several causes of UDP packet drops, but the primary reason is drops in the send/receive buffer. This is particularly true in the case of the loopback network

### What this PR changs

- Monitor UDP recv buffer errors and UDP send buffer errors

## Related links

None

## How was this PR tested?

- `total UDP rcv buf errors` and `total UDP snd buf errors` are monitored, and they are retrieved from `/proc/net/snmp`
- When errors per unit time is greater than threshold (>=1), status becomes warning
- The CPU usage of system_monitor doesn't change
  - Without this PR: 13.6% / 1600%
  - With this PR: 12.6% / 1600%
  - (the CPU usage of system_monitor in 16-core ECU)

image

### How to cause UDP buf errors

Run the following scripts

```sender.py
import socket
import time

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

# To cause snd buf error, enable the following line and run the following command
# $> sudo tc qdisc add dev lo root netem delay 100ms
# $> sudo tc qdisc del dev lo root    # fix setting
sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 4)

target_ip = "127.0.0.1"
target_port = 12345

while True:
    sock.sendto(b"a" * 1024, (target_ip, target_port))
    time.sleep(0.00001)
```

```receiver.py
import socket

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 1)
sock.bind(("0.0.0.0", 12345))

while True:
    try:
        data, addr = sock.recvfrom(1024)
    except Exception as e:
        print(f"Error: {e}")
```

## Notes for reviewers

None.

#### Additions and removals

None

#### Modifications

A new member named `net_monitor: UDP Buf Errors` is added into `/diagnostics` topic

#### Additions and removals

system/system_monitor/config/net_monitor.param.yaml

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| New     | `udp_buf_errors_check_duration` | `int` | `1`         | UDP buf errors check duration |
| New     | `udp_buf_errors_check_count` | `int` | `1`         | Generates warning when count of UDP buf errors during udp_buf_errors_check_duration reaches a specified value or higher |

#### Modifications

None

## Effects on system behavior

None.

### CODE REVIEW

**Comment by iwatake2222** [comment]
@ito-san @TetsuKawa 
Could you review this PR?
Also, please re-add `run-build-and-test-differential` label to run worflows (some workflows are stuck for some reasons...)

File: system/system_monitor/src/net_monitor/net_monitor.cpp
Comment (ito-san) [inline system/system_monitor/src/net_monitor/net_monitor.cpp:357]: I don't think fmt::format is necessary here.

File: system/system_monitor/src/net_monitor/net_monitor.cpp
Comment (ito-san) [inline system/system_monitor/src/net_monitor/net_monitor.cpp:368]: It looks like duplicate code, so please try to make it more reusable.

File: system/system_monitor/src/net_monitor/net_monitor.cpp
Comment (iwatake2222) [reply system/system_monitor/src/net_monitor/net_monitor.cpp]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/9538/commits/bfbbaac79edc8d02498b363e07098a4e0725ee12

**Comment by iwatake2222** [comment]
@ito-san @TetsuKawa 
Thank you for reviewing it. I updated the code as you suggested.
Could you review it again please?

File: system/system_monitor/src/net_monitor/net_monitor.cpp
Comment (TetsuKawa) [reply system/system_monitor/src/net_monitor/net_monitor.cpp:349]: @iwatake2222
I'm a bit concerned about the performance impact of reading /proc/net/snmp separately for each of these three items:

- IP fragmentation failed
- UDP send buffer errors
- UDP receive buffer errors

I think it would be better to consolidate these `std::ifstream ifs("/proc/net/snmp");` into a single function. (it is just an idea that creating a class including NetSnmp instances. the design is up to you.).

Comment (iwatake2222) [reply system/system_monitor/src/net_monitor/net_monitor.cpp:349]: @TetsuKawa 
Thank you for your feedback.  
I had exactly the same thought during the implementation.  

However, I concluded that there is no performance concern for the following reasons:
- generally, the overhead of accessing `/proc` is smaller than accessing a normal file
- the actual measurement results showed no noticeable increase in processing load

Of course, there might be a very very slight increase that is not visible at this level, but I judged that the current implementation is the better approach considering the following factors:
- Increased complexity of the implementation
- Higher maintenance costs if the number of monitored metrics increases in the future
- Increased risk of regression (with potential impacts on all `/proc/net/snmp`-related information)  

Therefore, I would like to keep the current design and implementation as it is :pray:

Comment (TetsuKawa) [reply system/system_monitor/src/net_monitor/net_monitor.cpp:349]: I understand. 
As you say, since /proc exists in memory rather than on SSD, it probably doesn't increase the process load much. (This is very educational for me.) 
I agree with your thinking!

**Comment by TetsuKawa** [comment]
LG**TM**
@ito-san 
I leave the rest to your judgment.

**Comment by iwatake2222** [comment]
@ito-san 
Friendly reminder...

**Comment by ito-san** [comment]
> @ito-san Friendly reminder...
Sorry, the codebase was different between this branch and my project, so my Autoware was broken after I merged this branch.
I've updated the branch, so please wait a little longer.

**Comment by ito-san** [comment]
Verified that no errors occur during startup
image

**Comment by ito-san** [comment]
It seems to be working well
image

**Review by ito-san — APPROVED** [review]
LGTM

**Comment by ito-san** [comment]
@iwatake2222 
Please fix this.
/__w/autoware.universe/autoware.universe/system/system_monitor/src/net_monitor/net_monitor.cpp:814:26: error: folding type 'unsigned int' into type 'int' might result in loss of precision [bugprone-fold-init-type,-warnings-as-errors]
  value_per_unit_time_ = std::accumulate(queue_.begin(), queue_.end(), 0);
                         ^

**Comment by iwatake2222** [comment]
@ito-san 
Could you add label to run workflow?

**Comment by iwatake2222** [comment]
@ito-san 
Thank you for reviewing it !!
I fixed the code and all required checks are passed (pre-commit-optional fails due to link error, but it's not related to this PR)

Could you merge this PR? I'm asking because `Only those with write access to this repository can merge pull requests.`

**Pull request merged**
